cmd/dp-recipe-api/main.go

```diff
@@ -37,86 +43,58 @@ func main() {
 	//Feature Flag for Mongo Connection
 	enableMongoData := cfg.MongoConfig.EnableMongoData
 
-	if enableMongoData {
-		mongodb := &mongo.Mongo{
-			Collection: cfg.MongoConfig.Collection,
-			Database:   cfg.MongoConfig.Database,
-			URI:        cfg.MongoConfig.BindAddr,
-		}
-
-		var err error
-		mongodb.Session, err = mongodb.Init()
-		if err != nil {
-			log.Event(ctx, "failed to initialise mongo", log.FATAL, log.Error(err))
-			os.Exit(1)
-		}
+	mongodb := &mongo.Mongo{
+		Collection: cfg.MongoConfig.Collection,
+		Database:   cfg.MongoConfig.Database,
+		URI:        cfg.MongoConfig.BindAddr,
 	}
 
-	router := mux.NewRouter()
-	router.Methods("GET").Path("/health").HandlerFunc(healthcheck)
-	router.Methods("GET").Path("/recipes").HandlerFunc(recipeListHandler)
-	router.Methods("GET").Path("/recipes/{id}").HandlerFunc(recipeHandler)
-
-	log.Event(ctx, "starting http server", log.INFO, log.Data{"bind_addr": cfg.BindAddr})
-	srv := server.New(cfg.BindAddr, router)
-	if err := srv.ListenAndServe(); err != nil {
-		log.Event(ctx, "error starting http server for API", log.FATAL, log.Error(err))
+	var err error
+	mongodb.Session, err = mongodb.Init()
+	if err != nil {
+		log.Event(ctx, "failed to initialise mongo", log.FATAL, log.Error(err))
 		os.Exit(1)
 	}
-}
 
-func healthcheck(w http.ResponseWriter, req *http.Request) {
-	// Set status to 200 OK
-	w.WriteHeader(200)
-}
-
-func recipeListHandler(w http.ResponseWriter, req *http.Request) {
-	list := &recipe.FullList
-	c := len(list.Items)
-	list.Count = c
-	list.TotalCount = c
-	list.Limit = c
+	if enableMongoData {
+		//Create RecipeAPI instance with Mongo in datastore
+		store := store.DataStore{Backend: RecipeAPIStore{mongodb}}
+		api.CreateAndInitialiseRecipeAPI(ctx, *cfg, store)
 
-	b, err := json.Marshal(list)
-	if err != nil {
-		log.Event(req.Context(), "error returned from json marshall", log.ERROR, log.Error(err))
-		w.WriteHeader(http.StatusInternalServerError)
-		return
+	} else {
+		//Create RecipeAPI instance with no datastore
+		api.CreateAndInitialiseRecipeAPI(ctx, *cfg, store.DataStore{Backend: nil})
 	}
 
-	w.Header().Set("Content-Type", "application/json")
-	w.Write(b)
-}
+	apiErrors := make(chan error, 1)
 
-func recipeHandler(w http.ResponseWriter, req *http.Request) {
-	vars := mux.Vars(req)
-	recipeID := vars["id"]
-	logD := log.Data{"recipe_id": recipeID}
+	// Gracefully shutdown the application closing any open resources.
+	gracefulShutdown := func() {
```

Comment:
I don't know if graceful shutdown is part of this PR, feel free to ignore this comment otherwise.

If we want to check the timeout we can create a separate go-routine here, which uses the shutdown context, and calls cancel() at the end (with a defer), then the main thread would wait for shutdownContext.Done(), and exit if there is no error.
An example would be the code in `dp-dataset-api`

Reply:
Different PR